src/services/services.ts

```diff
@@ -6409,35 +6436,45 @@ namespace ts {
-                                return [unionProperty];
-                            }
-                            else {
-                                const result: Symbol[] = [];
-                                forEach((<UnionType>contextualType).types, t => {
-                                    const symbol = t.getProperty(name);
-                                    if (symbol) {
-                                        result.push(symbol);
-                                    }
-                                });
-                                return result;
-                            }
+            function getNameFromObjectLiteralElement(node: ObjectLiteralElement) {
+                if (node.name.kind === SyntaxKind.ComputedPropertyName) {
+                    const nameExpression = (<ComputedPropertyName>node.name).expression;
+                    // treat computed property names where expression is string/numeric literal as just string/numeric literal
+                    if (isStringOrNumericLiteral(nameExpression.kind)) {
+                        return (<LiteralExpression>nameExpression).text;
+                    }
+                    return undefined;
+                }
+                return (<Identifier | LiteralExpression>node.name).text;
+            }
+
+            function getPropertySymbolsFromContextualType(node: ObjectLiteralElement): Symbol[] {
+                const objectLiteral = <ObjectLiteralExpression>node.parent;
+                const contextualType = typeChecker.getContextualType(objectLiteral);
+                const name = getNameFromObjectLiteralElement(node);
+                if (name && contextualType) {
+                    if (contextualType.flags & TypeFlags.Union) {
+                        // This is a union type, first see if the property we are looking for is a union property (i.e. exists in all types)
+                        // if not, search the constituent types for the property
+                        const unionProperty = contextualType.getProperty(name);
+                        if (unionProperty) {
+                            return [unionProperty];
                         }
                         else {
-                            const symbol = contextualType.getProperty(name);
-                            if (symbol) {
-                                return [symbol];
-                            }
+                            const result: Symbol[] = [];
+                            forEach((<UnionType>contextualType).types, t => {
+                                const symbol = t.getProperty(name);
+                                if (symbol) {
+                                    result.push(symbol);
+                                }
+                            });
+                            return result;
+                        }
+                    }
+                    else {
+                        const symbol = contextualType.getProperty(name);

```

Comment:
Minor: I think it may be easier to read if you move these three lines outside (because it is the first thing we will do regardless whether the contextualType is a union type or not. Then if it fails check if contextualType is a union type and look into each constituent.